### DISCUSSION TITLE
splitFasta by: variable

### DISCUSSION THREAD

**Discussion by JAMKuttan** [nextflow-io/nextflow#4137] [Q&A]
Hi I am using the nextflow scripting language and want to use the splitFasta function but split by a variable that forces a total of 50 files.

nFasta = (infasta.countFasta()/50).toInteger() + 1
infasta.splitFasta( by: $nFasta, file: true )

I am not sure how to access this value and pass it into the by statement.

**Comment by bentsherman** [comment]
Hi @JAMKuttan , since `countFasta()` returns a channel, you can't use it like a regular variable. You'll have to count the number of FASTA records with a custom function. You can follow [this example](https://nextflow.io/docs/edge/script.html#reading-a-file-line-by-line) to count the lines of a file.

**Comment by jpfeuffer** [comment]
Hi! I am trying to do the same, but I am not sure how the linked example helps.
Whenever I do:

```
count = 0
myFile.eachLine { str ->
    count++
}
``` 

`count` will not be updated outside. See also https://github.com/nextflow-io/nextflow/issues/1702 and https://github.com/nextflow-io/nextflow/discussions/3388

The workaround in the issue is also not really satisfying since it involves spawning a whole process just to compute something very basic like the size of a channel.

**Reply by bentsherman** [reply]
This example works for me:

```groovy
myFile = file('main.nf')
count = 0
myFile.eachLine {
    count++
}

println count
```

**Reply by jpfeuffer** [reply]
Hmm, you are correct. In that simple example, this works.
But try this:

```groovy
#!/usr/bin/env nextflow
nextflow.enable.dsl=2

process CREATEFILE {
    input:
    val s

    output:
    path "file.txt", emit: out

    script:
    """
    echo "$s" >> file.txt
    """
}

workflow CREATEFOO {
    take:
        s
    main:
        out = CREATEFILE(s).out
    emit:
        out
}

workflow MYEACHLINE {
    take:
        x

    main:
        c = 0
        x.eachLine{
            c++
        }
        println c
    emit:
        c
}

workflow {
    string = "foo"
    f = CREATEFOO(string).out
    MYEACHLINE(f)
}
```

This prints `0` for me.

**Reply by bentsherman** [reply]
Using `eachLine` doesn't really make sense in that example because you are using it on a channel instead of a file. I would put the counting code in a regular function rather than a workflow.

**Reply by jpfeuffer** [reply]
Thanks for your input! Could you give me an example? Or what do you mean by "function" exactly? A process?

**Reply by bentsherman** [reply]
```groovy
def myCountLines(file) {
    def count = 0
    file.eachLine{
        count++
    }
    return count
}
```

**Reply by jpfeuffer** [reply]
Ah I see, but in my case my file is indeed produced by a previous process and therefore comes out of a channel. Would your solution still work?

**Reply by bentsherman** [reply]
Right, so instead of `MYEACHLINE(f)` you would do `f.map { myCountLines(it) }` to get the lines in each file.

Although I should mention, since the original issue was about using `splitFasta` with a count based on the lines in the file, that you won't be able to do that if the file itself is coming from a channel. Because then the count is computed in a channel and you can't pull it out to pass it to the `by` option.

**Reply by bentsherman** [reply]
In fact, if you are doing `f.map { myCountLines(it) }` then you might as well do `f.countLines()` instead.
